### PR TITLE
[new feature] extend option

### PR DESCRIPTION
**Pull request by Andarist**
Fixes https://github.com/rollup/rollup/issues/827

This addresses inconsistent behaviour between `named` and `default` output. Ideally whenever possible IIFE should output a simple variable, not extending a global object.

As extending global still has its use cases this PR allows for that with a new option - `extend`, also 'namespaced' modules are preserving the old behaviour, as it is not possible to achieve a simple variable for them anyway.

When implementing this Ive also realised that the same thing should be implemented for `umd` and I can do that, aint sure if that should be a matter of another PR or this one. **NOTE:** umd change would affect d3, its however easily fixable - first of all we can add a warning about the changed behaviour, second of all we can target d3 specifically and add `extend` option for them automatically (for a 1 minor version ofc)

Also I was not sure if I should maybe throw when validating the options object - `extend` now is only implemented for `iife` case, and for other module formats like `es` it doesnt even make sense, so maybe the user should be warned that he uses obsolete option.

Didn't yet include any docs about this, can do that. However would prefer to add this at the end, when PR gets approved.

cc @bfred-it @TrySound @mbostock @Victorystick

### CODE REVIEW

**Comment by bfred-it** [comment]
> the same thing should be implemented for `umd` and I can do that, aint sure if that should be a matter of another PR or this one

They should probably both behave the same way with globals, so it's best to also change UMD's behavior.

> we can target d3 specifically and add `extend` option for them automatically

I'd rather not have special _behavior_ like that for others. A special _warning_ would be enough. 

> for other module formats like `es` it doesnt even make sense

Right, `moduleName` and `extend` only apply to the IIFE bundle.

**Comment by Andarist** [comment]
> They should probably both behave the same way with globals, so it's best to also change UMD's behavior.

Ok. I'll work on that one soon too, will just add commits to this PR.

**Comment by Victorystick** [comment]
Nice work, but it doesn't look quite ready yet. Take the following trivial example:

```js
export var a = 1;
```

The resulting bundle would leave `myBundle` without a value, since none is returned by the function. You need to end the bundle with a `return exports;`.
```js
var myBundle = (function (exports) {
	'use strict';

	exports.a = 1;
}({}));
```

**Comment by Andarist** [comment]
oh damn, didnt spotted this one! gonna fix it ofc, thanks for the review

**Comment by Andarist** [comment]
Ok, the bug spotted by @Victorystick is fixed. UMD tweaks are also done.

While working on this though I've spotted that [this](https://github.com/rollup/rollup/blob/ef59ea2d8979570aeca6c434986ef1504697fce9/test/form/umd-noconflict/_expected/umd.js#L6) is not working properly. It suffers from the same problem that @Victorystick pointed out to me - nothing is returned from the closure. 

So I fixed that issue too.

Please review the PR and let me know if there is anything else I can do here. I would personally add some warnings (maybe even errors?) for options combinations that are not supposed to work together, such as:
- extend + format other than umd or iife
- umd + noConflict + exportMode: 'none'
Aint sure where such things are usually validated though.

**Comment by TrySound** [comment]
@Andarist Did you see my review?

**Comment by Andarist** [comment]
@TrySound unfortunately not, maybe you have forgotten to send it? I often forget that this new~ review tool doesnt send comments immediately

**Comment by TrySound** [comment]
You are right :)

File: src/utils/identifier-helpers.js [hunk 1]
```diff
@@ -6,13 +6,26 @@ const builtins = 'Infinity NaN undefined null true false eval uneval isFinite is
...
 	str = str
 		.replace( /-(\w)/g, ( _, letter ) => letter.toUpperCase() )
-		.replace( /[^$_a-zA-Z0-9]/g, '_' );
+		.replace( illegalCharacters, '_' );
 
-	if ( /\d/.test( str[0] ) || blacklisted[ str ] ) str = `_${str}`;
+	if ( startsWithDigit(str) || blacklisted[ str ] ) str = `_${str}`;
 
 	return str;
 }
```
Comment (Andarist) [inline src/utils/identifier-helpers.js]: k, should I maybe rename [map-helpers](https://github.com/rollup/rollup/blob/ef59ea2d8979570aeca6c434986ef1504697fce9/src/utils/map-helpers.js) too?

Comment (TrySound) [inline src/utils/identifier-helpers.js]: Would be great :)

**Comment by Andarist** [comment]
@TrySound I've adjusted the code with your suggestions. Please take a look.

**Comment by Rich-Harris** [comment]
thanks! releasing as 0.44

**Comment by dnoyes** [comment]
@Rich-Harris @Andarist is there somewhere I can go to see example usage?

**Comment by Andarist** [comment]
example usage is just passing `extend: true` to the config options, unfortunately I didnt test it properly with CLI usage and its not working when passed from CLI arguments, fix is on the way though - https://github.com/rollup/rollup/pull/1482

**Pull request merged**